cpp/src/groupby/hash/groupby.cu

```diff
@@ -421,36 +425,31 @@ void sparse_to_dense_results(table_view const& keys,
  * @brief Construct hash map that uses row comparator and row hasher on
  * `d_keys` table and stores indices
  */
-auto create_hash_map(table_device_view const& d_keys,
-                     bool keys_have_nulls,
-                     null_policy include_null_keys,
-                     rmm::cuda_stream_view stream)
+auto create_hash_map(
+  std::shared_ptr<cudf::experimental::row::hash::preprocessed_table> preprocessed_keys,
+  cudf::size_type const num_keys,
+  null_policy const include_null_keys,
+  rmm::cuda_stream_view stream)
 {
   size_type constexpr unused_key{std::numeric_limits<size_type>::max()};
   size_type constexpr unused_value{std::numeric_limits<size_type>::max()};
 
-  using map_type =
-    concurrent_unordered_map<size_type,
-                             size_type,
-                             row_hasher<cudf::detail::default_hash, nullate::DYNAMIC>,
-                             row_equality_comparator<nullate::DYNAMIC>>;
-
   using allocator_type = typename map_type::allocator_type;
 
   auto const null_keys_are_equal =
     include_null_keys == null_policy::INCLUDE ? null_equality::EQUAL : null_equality::UNEQUAL;
+  auto const has_null = nullate::DYNAMIC{true};  // always has null TODO: why??
```

Comment:
How was this being initialized other than as always `true`? It looks like the `keys_have_nulls` parameter is no longer passed to this function, so how was it otherwise being computed?

Reply:
So I've fixed the lifetime bug and it still needs to be `true` instead of `keys_have_nulls`. I suspect it has something to do with https://github.com/rapidsai/cudf/blob/branch-22.06/cpp/src/groupby/groupby.cu#L79, where the flattened columns are set to always have nulls as well. @mythrocks Do you recall why the flattened columns always generate a null column?

Reply:
I think I find the reason. Let's say we have a table of one `STRUCT` column and none of the `STRUCT` is null itself:
```
  +-----------------+
  | s1{s2{a,b},  c} |
  +-----------------+
0 |  { {1, 1}, "a"} |
1 |  { {1, 2}, "b"} |
2 |  {   null, "c"} |
  +---------------- +
```
When row 2 has a null element in the struct, `cudf::has_nulls(table)` still returns `false`. This becomes an issue for the follow-up row hasher/comparator since what they really expect is whether structs have null elements as opposed to whether there are null structs in the column. `cudf::has_nulls()` obviously returns the later.

Reply:
That issue was mitigated. I added `has_nested_null` (in `table_view.hpp` I believe) so you can check null at all levels.

Reply:
@ttnghia Exactly what I'm looking for. Thanks!

I think we should update `row_hasher.device_hasher()` and `self_comparator.device_comparator()` APIs to not take `nullate` argument since it can be initialized when constructing `row_hasher`/`self_comparator`.

https://github.com/rapidsai/cudf/blob/branch-22.06/cpp/src/stream_compaction/distinct.cu#L72 seems buggy since it uses `has_nulls` instead of `has_nested_nulls` as well.

cc @devavret

Reply:
> I think we should update row_hasher.device_hasher() and self_comparator.device_comparator() APIs to not take nullate argument since it can be initialized when constructing row_hasher/self_comparator.

That's how it was until a review requesting it to change.

> https://github.com/rapidsai/cudf/blob/branch-22.06/cpp/src/stream_compaction/distinct.cu#L72 seems buggy since it uses has_nulls instead of has_nested_nulls as well

I think the behaviour of has_nulls should be changed to that of has_nested_nulls.

Reply:
> That's how it was until a review requesting it to change.

OK. found it https://github.com/rapidsai/cudf/pull/10164#discussion_r829113477.

> I think the behaviour of has_nulls should be changed to that of has_nested_nulls.

Right, not sure why we are keeping both.

Reply:
We have both because they mean different things.

`has_nulls()` is just a convenience for `null_count() > 0` (similar to `is_empty()` vs `size() > 0`).

That obviously means something very different from `has_nested_nulls()`.

Reply:
> think we should update row_hasher.device_hasher() and self_comparator.device_comparator() APIs to not take nullate argument since it can be initialized when constructing row_hasher/self_comparator.
> That's how it was until a review requesting it to change.

Did I request that? I don't remember that.

It seems like the `Nullate` should be specified at the construction of the owning comparator type. Specifying it in the `device_comparator`  just invites the opportunity for passing incorrect or different values from one call to the next.

Reply:
> It seems like the Nullate should be specified at the construction of the owning comparator type.

I did it so that if you wanted dynamic then you only had to specify it as template parameter and the has nulls check would be done by the owning type because it could just check the table view. There was no need to pass any nullate argument. 

Another reason I changed it was because the has_nulls check causes kernel launches and if the user already has those at hand then we avoid those.

Since the owning type doesn't care about nulls during the preprocessing, I didn't add it. With a template owning type, the construction would have to be in the if block that branches on has_nulls. Causing unnecessary duplicate code. 

> Specifying it in the device_comparator just invites the opportunity for passing incorrect or different values from one call to the next.

This is allowed. It's not a bug.

Reply:
> https://github.com/rapidsai/cudf/blob/branch-22.06/cpp/src/stream_compaction/distinct.cu#L72 seems buggy since it uses has_nulls instead of has_nested_nulls as well.

Created #10848 to fix the bug in `distinct`. Another thing I want to mention is that the requirement of `Nullate` is **not** consistent between `row_hasher` and `self_comparator` i.e. with `nullable = has_nulls(table)`: 
```
  @ indicates null
  +-----------------+
  | s1{s2{a,b},  c} |
  +-----------------+
0 |  { {1, 1}, "a"} |
1 |  {@{1, 1}, "a"} |
  +---------------- +
```
`self_comparator` will see row 0 and row 1 as equal but `row_hasher` returns different hash values. 

As a result:

- `self_comparator` must be used with `has_nested_nulls`
- `row_hasher` can work with either `has_nested_nulls` or `has_nulls`

Reply:
I will leave the current PR as it is and file an issue to track the potential hasher bug.